### ISSUE TITLE
text with control codes

### ISSUE THREAD

**Issue by Midnighter** [pyparsing/pyparsing#274]
Hi,

I have a piece of text (from a source outside of my control) which contains [control codes](https://en.wikipedia.org/wiki/C0_and_C1_control_codes), for example, the non-breaking separator NBSP (`0xA0`). I have added the codes where that makes sense to the list of default whitespace characters.

```py
from pyparsing import ParserElement

ParserElement.setDefaultWhitespaceChars(" \t\n\r" + "\xa0")
```

This works well except for cases where a control code appears inside of a `nestedExpr`. Do I need to adjust the whitespace characters for that differently?

**Comment by ptmcg** [comment]
Can you post an example showing this in more detail?

**Comment by Midnighter** [comment]
I'll dig out a minimal example. In the meantime, I've switched to pre-processing the entire document and removing the control codes. It made my life a lot easier.

**Comment by Midnighter** [comment]
Okay, I have come up with a small example that is similar to my other question.

```py
import pyparsing as pp

pp.ParserElement.setDefaultWhitespaceChars(" \t\n\r" + "\x02")

integer = pp.pyparsing_common.integer
tags = ("#" + pp.delimitedList(integer)("tags") + "#")
name = pp.Word(pp.alphas, pp.printables)
references = ("<" + pp.delimitedList(integer)("references") + ">")

expr = pp.Group(
    tags
    + name("name")
    + pp.nestedExpr()("comment")
    + references
)

sample = """#4# More   (residues 16 236, designated as 1510-N) <1,2,3,4,6,7,8>"""

print("Sample with normal space.")
print(expr.parseString(sample, parseAll=True)[0].dump())

sample2 = """#4# More   (residues 16\x02236, designated as 1510-N) <1,2,3,4,6,7,8>"""

print("Sample with non-breaking space control code.")
print(expr.parseString(sample2, parseAll=True)[0].dump())
```

I expected the result from `sample` and `sample2` to be the same. However, I get the following:

```
['#', 4, '#', 'More', ['residues', '16', '236,', 'designated', 'as', '1510-N'], '<', 1, 2, 3, 4, 6, 7, 8, '>']
- comment: [['residues', '16', '236,', 'designated', 'as', '1510-N']]
  [0]:
    ['residues', '16', '236,', 'designated', 'as', '1510-N']
- name: 'More'
- references: [1, 2, 3, 4, 6, 7, 8]
- tags: [4]
Sample with non-breaking space control code.
Traceback (most recent call last):
  File "/home/moritz/Codebase/Midnighter/BRENDA-Parser/mre.py", line 25, in <module>
    print(expr.parseString(sample2, parseAll=True)[0].dump())
  File "/home/moritz/.pyenv/versions/brenda/lib/python3.8/site-packages/pyparsing.py", line 1955, in parseString
    raise exc
  File "/home/moritz/.pyenv/versions/brenda/lib/python3.8/site-packages/pyparsing.py", line 2899, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pyparsing.ParseException: Expected ")", found '2'  (at char 24), (line:1, col:25)
```

**Comment by ptmcg** [comment]
The problem is that the default content expression for `nestedExpr()` is created *before* your custom definition of the default whitespace characters.

This is fixed in version 3.0, which resets the whitespace characters for all pre-defined expressions in whenever you call `setDefaultWhitespaceChars()`.

In 2.4.7, you can workaround the issue by defining your own custom content expression. Your custom content will use the updated whitespace characters:

```
content_word = pp.Word(pp.printables, excludeChars="(),")
content = (pp.delimitedList(content_word))
expr = pp.Group(
    tags
    + name("name")
    + pp.nestedExpr(content=content)("comment")
    + references
)

```

**Comment by Midnighter** [comment]
Thank you for the explanation and fix.
